src/Chapter3.hs

```diff
@@ -1,3 +1,7 @@
+{-# LANGUAGE GADTs #-}
+-- NOTE: I don't fully know what all of these do... I just know that I wanted to put constraints in places that GHC
+-- didn't like unless I added these extensions. Is there a performance cost? Some other hidden cost?
```

Comment:
Woops, didn't mean to leave that in. It is a leftover from when I was experimenting with many different possible solution designs.

It was before I had `Fightable` and wanted to do something like `data Fight = (Fighter a, Fighter b) => (a, b)`. But I found that made my solution very brittle because each combination of different types of fighters would be a different type and require a different implementation.

Reply:
Oh, I see now!

Yeah, I see the problem. It is not that easy to work with such types, but I think in this case, the type could be replaced with the function with constraints :
```haskell
fight :: (Fighter a, Fighter b) => a -> b -> Result
```